// jamai/src/resources/llm/chat.ts
import { z } from "zod";

/**
 * Represents who said a chat message.
 */

const ChatRoleSchema = z.enum(["system", "user", "assistant", "function"]);

export const ChatEntrySchema = z.object({
    role: ChatRoleSchema,
    content: z.string(),
    name: z.string().optional().nullable(),
});

const RAGParamsSchema = z.object({
    search_query: z.string().optional(),
    k: z.number().optional(),
    fetch_k: z.number().optional(),
    document_ids: z.array(z.string()).optional(),
    rank_profile: z.enum(["bm25", "semantic", "hybrid", "hybrid_log"]).optional(),
    rerank: z.boolean().optional(),
    concat_reranker_input: z.boolean().optional(),
});

const FunctionSpecSchema = z.object({
    name: z.string(),
    description: z.string().optional(),
    parameters: z.record(z.string(), z.any()),
});

const ToolSpecSchema = z.object({
    type: z.string().optional(),
    function: FunctionSpecSchema,
});

const FunctionChoiceSpecSchema = z.object({
    name: z.string(),
});

const ChatCompletionUsageSchema = z.object({
    prompt_tokens: z.number().nullable().optional(),
    completions_tokens: z.number().nullable().optional(),
    total_tokens: z.number().nullable().optional(),
});

/**
 * Represents a message in the chat context.
 */
// export type ChatEntry = {
//   role: ChatRole;
//   content: string;
//   name?: string | null;
// };

export const ChatRequestSchema = z.object({
    id: z.string().optional(),
    model: z.string().optional(),
    messages: z.array(ChatEntrySchema),
    rag_params: RAGParamsSchema.nullable().optional(),
    tools: z.array(ToolSpecSchema).nullable().optional(),
    tool_choice: z.union([z.string(), ToolSpecSchema]).nullable().optional(),
    temperature: z.number().optional().default(1.0).optional(),
    top_p: z.number().optional().default(1.0).optional(),
    n: z.number().optional().default(1).optional(),
    // stream: z.boolean().optional().default(true),
    stop: z.array(z.string()).optional(),
    max_tokens: z.number().default(2048).optional(),
    presence_penalty: z.number().optional().default(0.0).optional(),
    frequency_penalty: z.number().optional().default(0.0).optional(),
    logit_bias: z.record(z.string(), z.any()).default({}).optional(),
    user: z.string().default("").optional(),
});

export const ChatCompletionChoiceSchema = z.object({
    message: ChatEntrySchema,
    index: z.number(),
    finish_reason: z.string().nullable(),
});

export const ChatCompletionChoiceDeltaSchema = z.object({
    message: ChatEntrySchema,
    index: z.number(),
    finish_reason: z.string().nullable(),
    delta: ChatEntrySchema.nullable(),
});

export const ReferencesSchema = z.object({
    object: z.string(),
});

export const ChatCompletionChunkSchema = z.object({
    id: z.string(),
    object: z.enum(["chat.completion", "chat.completion.chunk"]),
    created: z.number(),
    model: z.string(),
    usage: ChatCompletionUsageSchema.optional(),
    // choices: z.array(ChatCompletionChoiceSchema),
    choices: z.union([z.array(ChatCompletionChoiceSchema), z.array(ChatCompletionChoiceDeltaSchema)]),
    references: ReferencesSchema.nullable().optional(),
});

export const StreamChatCompletionChunkSchema = z.object({
    id: z.string(),
    object: z.enum(["chat.completion.chunk"]),
    created: z.number(),
    model: z.string(),
    usage: ChatCompletionUsageSchema,
    choices: z.array(ChatCompletionChoiceDeltaSchema),
    // choices: z.union([z.array(ChatCompletionChoiceSchema), z.array(ChatCompletionChoiceDeltaSchema)]),
    references: ReferencesSchema.nullable().optional(),
});

export type ChatRole = z.infer<typeof ChatRoleSchema>;
export type ChatEntry = z.infer<typeof ChatEntrySchema>;
export type RAGParams = z.infer<typeof RAGParamsSchema>;
export type FunctionSpec = z.infer<typeof FunctionSpecSchema>;
export type ToolSpec = z.infer<typeof ToolSpecSchema>;
export type FunctionChoiceSpec = z.infer<typeof FunctionChoiceSpecSchema>;
export type ChatCompletionUsage = z.infer<typeof ChatCompletionUsageSchema>;
export type ChatRequest = z.infer<typeof ChatRequestSchema>;
export type References = z.infer<typeof ReferencesSchema>;
export type ChatCompletionChoice = z.infer<typeof ChatCompletionChoiceSchema>;
export type ChatCompletionChoiceDelta = z.infer<typeof ChatCompletionChoiceDeltaSchema>;
export type ChatCompletionChunk = z.infer<typeof ChatCompletionChunkSchema>;
export type StreamChatCompletionChunk = z.infer<typeof StreamChatCompletionChunkSchema>;

// export type ChatRequest = {
//     id?: string;
//     model?: string;
//     messages?: Array<ChatEntry>;
//     rag_params?: RAGParams | null;
//     tools?: Array<ToolSpec> | null;
//     tool_choice?: string | ToolChoiceSpec | null;
//     temperature?: number;
//     top_p?: number;
//     n?: number;
//     stream?: boolean;
//     stop?: Array<string>;
//     max_tokens?: number;
//     presence_penalty?: number;
//     frequency_penalty?: number;
//     logit_bias?: {
//       [key: string]: unknown;
//     };
//     user?: string;
// };

// /**
//  * Represents who said a chat message.
//  */
// export type ChatRole = "system" | "user" | "assistant" | "function";

// export type RAGParams = {
//   search_query?: string;
//   k?: number;
//   fetch_k?: number;
//   access_level?: number;
//   document_ids?: Array<string>;
//   rank_profile?: "bm25" | "semantic" | "hybrid" | "hybrid_log";
//   rerank?: boolean;
//   concat_reranker_input?: boolean;
// };

// export type ToolSpec = {
//   type?: string;
//   function: FunctionSpec;
// };

// export interface ToolChoiceSpec {
//   type: string;
//   function: FunctionChoiceSpec;
// }

// export interface FunctionSpec {
//   name: string;
//   description?: string;
//   parameters: { [key: string]: any };
// }

// export interface FunctionChoiceSpec {
//   name: string;
// }

// export interface ChatCompletionUsage {
//     prompt_tokens: number;
//     completions_tokens: number;
//     total_tokens: number;
// }

// export interface ChatCompletionChoice {
//     message: ChatEntry;
//     index: number;
//     finish_reason: string;
// }

// export interface ChatCompletionChoiceDelta {
//     message: ChatEntry;
//     index: number;
//     finish_reason: string;
//     delta: ChatEntry;
// }

// export interface References {
//     object: string;
// }

// export interface ChatCompletionChunk {
//     id: string;
//     object: string;
//     created: number;
//     model: string;
//     usage: ChatCompletionUsage;
//     choices: Array<ChatCompletionChoice> | Array<ChatCompletionChoiceDelta>;
//     references: References | null | undefined;

// }
